\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={Analyzing ChIP-seq data},
            pdfauthor={Mireia Ramos-Rodriguez},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{longtable,booktabs}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{Analyzing ChIP-seq data}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{Mireia Ramos-Rodriguez}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{March 14, 2019}

% Using xcolors with dvipsnames before mdframed to avoid option clash
\usepackage{xcolor}
\usepackage[framemethod=TikZ]{mdframed}

% Defining the base style of the box as bbox
\mdfdefinestyle{bbox}{
  leftmargin = 0cm,%
  rightmargin = 0cm,%
  innerleftmargin = 0.5cm,%
  innerrightmargin = 0.5cm,%
  roundcorner = 5pt,%
  linewidth = 2pt,%
  nobreak=false%,
  %needspace=1in % Trying to avoid a page break after the title but does not do the trick...
}

% Defining custom box environment (cbox) with 2 arguments:
% The first optional (defaults to empty string) is the title of the box
% The second (mandatory) defines the color of the box (linecolor and background is linecolor!30)

% body: fill (#2), colour (#3)
% header: fill (#4), colour (#5)
\newenvironment{cbox}[5][]{
  \def\temp{#1}\ifx\temp\empty
    \begin{mdframed}[style=bbox,%
      linecolor = #4,%
      fontcolor = #3,%
      innertopmargin = 10pt,%
      skipabove = 10pt,%
      backgroundcolor = #2]%
  \else
    \begin{mdframed}[style=bbox,%
      linecolor = #4,%
      fontcolor = #3,%
      backgroundcolor = #2,%
      innertopmargin = 3pt,%
      skipabove = 10pt,%
      frametitle={\tikz[baseline = (current bounding box.east), outer sep = 0pt]
                  \node[anchor = east, rounded corners = 3pt, rectangle, color = #5, fill = #4]{#1};},%
                        frametitleaboveskip = \dimexpr-\ht\strutbox\relax]
  \fi
}{\end{mdframed}}

\mdfdefinestyle{input}{
  frametitle = {},%
  leftmargin = 0cm,%
  rightmargin = 0cm,%
  innerleftmargin = 0.2cm,%
  innerrightmargin = 0.2cm,%
  innertopmargin=0.5em,%
  roundcorner = 2pt,%
  linecolor = lightgray,%
  linewidth = 1pt,%
  backgroundcolor = lightgray!30
}

% Defining cbox macro to draw a coloured block
% around the content

\newcommand{\cboxs}[5][]{\begin{cbox}[#1]{#2}{#3}{#4}{#5}}
\newcommand{\cboxe}{\end{cbox}}

\makeatletter

\begin{document}
\maketitle

\section{Introduction to ChIP-seq}\label{introduction-to-chip-seq}

\subsection{What is ChIP-seq?}\label{what-is-chip-seq}

Chromatin immunoprecipitation experiments followed by sequencing
(ChIP--seq) detect \textbf{protein--DNA binding} events and chemical
\textbf{modifications of histone} proteins (\emph{Figure 1}).

\begin{figure}

{\centering \includegraphics[width=2.63in]{img/protocol_chipseq} 

}

\caption{Comparison of ChIP-seq experimental protocols (adapted from Furey 2012, Nature Reviews).}\label{fig:protocol}
\end{figure}

Profiles obtained from transcription factor (TF) ChIP-seq and histone
ChIP-seq are different (\emph{Figure 2}):

\begin{itemize}
\tightlist
\item
  TF ChIP-seq detects enrichment at the \textbf{cis-regulatory element}
  (CRE), where the TFs bind.
\item
  Histone ChIP-seq detects enrichment at the \textbf{sorrounding
  nucleosomes}, where histone residues are modified.
\end{itemize}

\begin{figure}

{\centering \includegraphics[width=3.6in]{img/tf_histone_cres} 

}

\caption{The transcription factor and histone modification landscape for inactive and active cis-regulatory elements (CRE; promoters and enhancers) and the corresponding read profiles (from Pundhir et al. 2015, Frontiers in Genetics).}\label{fig:landscape}
\end{figure}

\subsection{Sequencing parameters}\label{sequencing-parameters}

You have to consider that when sequencing ChIP-seq, many of the
parameters you can choose depend on the specifics of your experiment,
your question and your budget! However, to provide you with some general
guidelines, this are some general recommendations:

\begin{itemize}
\tightlist
\item
  \textbf{Paired or single-end}. Usually single-end reads is sufficient
  for ChIP-seq analysis.
\item
  \textbf{Sequencing depth}. Between 30M (for narrow peaks) and 60M (for
  broad marks) sequenced reads should yield good profiles.
\item
  \textbf{Read length}. 50bp reads are usually sufficient. Could be
  interesting to make them longer when dealing with broader regions, but
  not that necessary when regions are narrow.
\end{itemize}

\section{ChIP-seq Workflow}\label{chip-seq-workflow}

\subsection{The dataset}\label{the-dataset}

The datasets we are going to use for this workshop are from
\href{https://www.ncbi.nlm.nih.gov/pubmed/30136421}{Hlady et al., 2019
\emph{Hepatology}}, in which they compare several assays in normal liver
with cirrotic liver and hepatocellular carcinoma (HCC) samples from
human donors.

All raw data from their assays can be accessed via their GEO ID:
\textbf{\href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112221}{GSE112221}}.
The \textbf{\href{https://www.ncbi.nlm.nih.gov/geo/}{Gene Expression
Omnibus (GEO)}} is a public functional genomics data repository that
accepts array- and sequence-based data. Each dataset is given a GEO ID
that usually starts with \texttt{GSE}. Inside each dataset, you can find
independent samples with identifiers prefixed \texttt{GSM}.

In GEO you can download already processed data, such as bed files.
However, we here are interested in downloading the raw fastq files so we
can do all the processing ourselves. Therefore, for each sample we are
interested in, we need to find the SRA ID (\texttt{SRX}, you can find it
in the sample page in GEO) and from there, gather the run ID
(\texttt{SRR}) which will give us access to the raw data.

\definecolor{unnamed-chunk-1.body.fill}{RGB}{221, 160, 221}
\definecolor{unnamed-chunk-1.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-1.header.fill}{RGB}{177, 128, 177}
\definecolor{unnamed-chunk-1.header.colour}{RGB}{0, 0, 0}

\cboxs[Exercise]{unnamed-chunk-1.body.fill!100}{unnamed-chunk-1.body.colour!100}{unnamed-chunk-1.header.fill!100}{unnamed-chunk-1.header.colour!100}

Go to the dataset website at GEO
(\href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112221}{here}).
Look around and, using what is explained above, find the \textbf{run ID}
for the normal liver sample of the H3K27ac ChIP-seq.

\cboxe

\definecolor{unnamed-chunk-2.body.fill}{RGB}{172, 255, 175}
\definecolor{unnamed-chunk-2.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-2.header.fill}{RGB}{138, 204, 140}
\definecolor{unnamed-chunk-2.header.colour}{RGB}{0, 0, 0}

\cboxs[Look at the result]{unnamed-chunk-2.body.fill!30}{unnamed-chunk-2.body.colour!100}{unnamed-chunk-2.header.fill!30}{unnamed-chunk-2.header.colour!100}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Go to the section \textbf{Samples (55)} and click on \texttt{+\ More}.
\item
  Find the sample named \textbf{NL1\_h3k27ac\_h3k27ac\_chipseq} and
  click on the GEO sample id
  (\href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3061124}{GSM3061124}).
\item
  Find the section named \textbf{Relations} and then click on the SRA ID
  (\href{https://www.ncbi.nlm.nih.gov/sra?term=SRX3832997}{SRX3832997})
\item
  On the bottom you will see a table named \textbf{Runs} that includes
  all the runs for that sample.
\item
  The \textbf{run ID} for the normal liver H3K27ac sample is
  \textbf{SRR6880492}
\end{enumerate}

\cboxe

You can download the files directly from the SRA website by clicking on
the run ID, but you can also do so from the terminal using the
\textbf{\href{https://www.ncbi.nlm.nih.gov/sra/docs/toolkitsoft/}{SRA
toolkit}} tool \texttt{fastq-dump}.

With \texttt{fastq-dump} you will be able to download fastq files
directly from SRA. Since these samples are paired-end, you need to
include the argument \texttt{-\/-split-3} to send each pair to a
different file. \texttt{-\/-gzip} will compress the fastq output to
fastq.gz.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Download fastq files from SRA database}
\ExtensionTok{fastq-dump}\NormalTok{ --split-3 --gzip -O your_folder/ SRR6880492 }
\end{Highlighting}
\end{Shaded}

Here you have a complete list of the samples we are going to use in this
workshop and their SRA run IDs:

\begin{longtable}[]{@{}lll@{}}
\toprule
\begin{minipage}[b]{0.32\columnwidth}\raggedright\strut
Sample names\strut
\end{minipage} & \begin{minipage}[b]{0.27\columnwidth}\raggedright\strut
SRA run ID\strut
\end{minipage} & \begin{minipage}[b]{0.32\columnwidth}\raggedright\strut
Description\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
NL1\_h3k27ac\_h3k27ac\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880492\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
H3K27ac ChIP-seq in Normal Liver\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Cirr1\_h3k27ac\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880493\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
H3K27ac ChIP-seq in Cirrotic Liver (Patient 1)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
HCC1\_h3k27ac\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880494\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
H3K27ac ChIP-seq in Hepatocellular Carcinoma (Patient 1)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Cirr3\_h3k27ac\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880495\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
H3K27ac ChIP-seq in Cirrotic Liver (Patient 3)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
HCC3\_h3k27ac\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880496\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
H3K27ac ChIP-seq in Hepatocellular Carcinoma (Patient 3)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
NL1\_input\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880507, SRR6880508\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Input ChIP-seq in Normal Liver\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Cirr1\_input\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880509, SRR6880510\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Input ChIP-seq in Cirrotic Liver (Patient 1)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
HCC1\_input\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880511, SRR6880512\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Input ChIP-seq in Hepatocellular Carcinoma (Patient 1)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Cirr3\_input\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880513, SRR6880514\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Input ChIP-seq in Cirrotic Liver (Patient 3)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
HCC3\_input\_chipseq\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright\strut
SRR6880515, SRR6880516\strut
\end{minipage} & \begin{minipage}[t]{0.32\columnwidth}\raggedright\strut
Input ChIP-seq in Hepatocellular Carcinoma (Patient 3)\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

\subsection{General workflow}\label{general-workflow}

A general workflow for ChIP-seq analysis consists on the following
steps:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  \textbf{Qualiy control}. Check if reads have good overall quality.
  Trim adapters and reads if approppriate.
\item
  \textbf{Alignment}. Align reads to a reference genome to obtain their
  genomic coordiantes.
\item
  \textbf{Post-processing}. Remove duplicates and unwanted chromosomes.
\item
  \textbf{Peak calling}. Find regions of enrichment (TF binding sites or
  nucleosomes with the histone mark of interest).
\item
  \textbf{Visualization}. Visualize profile in a genome browser.
\end{enumerate}

\subsection{Some technical details}\label{some-technical-details}

\section{Quality control}\label{quality-control}

Before starting any processing steps, one should check the overall
quality of the reads obtained from the sequencing facility. To do so, we
are going to use
\textbf{\href{https://www.bioinformatics.babraham.ac.uk/projects/fastqc/}{FastQC}},
a quality control tool for high-troughput sequencing data.

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{fastqc}\NormalTok{ folder_with_fastq/sample_reads.fastq}
\end{Highlighting}
\end{Shaded}

After checking that the overall quality is good, one can trim reads
and/or filter out reads with low overall quality using different
tools\footnote{We will not cover this in this workshop but if you are
  interested in this, you should check
  \href{https://github.com/OpenGene/AfterQC}{AfterQC}}.

\definecolor{unnamed-chunk-5.body.fill}{RGB}{221, 160, 221}
\definecolor{unnamed-chunk-5.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-5.header.fill}{RGB}{177, 128, 177}
\definecolor{unnamed-chunk-5.header.colour}{RGB}{0, 0, 0}

\cboxs[Exercise]{unnamed-chunk-5.body.fill!100}{unnamed-chunk-5.body.colour!100}{unnamed-chunk-5.header.fill!100}{unnamed-chunk-5.header.colour!100}

Now run \texttt{fastqc} for the control liver sample.

\cboxe

\definecolor{unnamed-chunk-6.body.fill}{RGB}{172, 255, 175}
\definecolor{unnamed-chunk-6.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-6.header.fill}{RGB}{138, 204, 140}
\definecolor{unnamed-chunk-6.header.colour}{RGB}{0, 0, 0}

\cboxs[Look at the result]{unnamed-chunk-6.body.fill!30}{unnamed-chunk-6.body.colour!100}{unnamed-chunk-6.header.fill!30}{unnamed-chunk-6.header.colour!100}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{fastqc}\NormalTok{ fastq/NL1_H3K27ac.fastq}
\end{Highlighting}
\end{Shaded}

\cboxe

\section{Alignment}\label{alignment}

We call \textbf{alignment} (or mapping) to the process of comparing
reads to a reference genome, scoring the similarity and then assigning
most likely genomic coordinates to each read.

\subsection{Reference genomes}\label{reference-genomes}

A \textbf{reference genome} is a database of DNA sequences, assembled as
a representative of a species genome. Usually, one can find two versions
of the reference genome: \emph{NCBI} (starts with \texttt{GRC}) and
\emph{UCSC} (\texttt{hg} for human, \texttt{mm} for mice). The
difference between the two basically lies in the naming of the
chromosomes: NCBI uses numbers (i.e.~1, 2, X) and UCSC numbers plus
\texttt{chr} (i.e.~chr1, chr2, chrX).

Nowadays, the most used versions of the human genome are the following:

\begin{itemize}
\tightlist
\item
  \textbf{hg19}/\textbf{GRCh37}. Even though this is not the latest
  version, is the most broadly used. If you do not specifically need
  high accuracy when mapping your reads, you should use this, as many
  tools (such as \href{http://great.stanford.edu/public/html/}{GREAT})
  will only work with this version.
\item
  \textbf{hg38}/\textbf{GRch38}. This is the newest version. Even though
  it was relased in 2013, many labs still use the previous version.
\end{itemize}

There are tools to convert coordinates from one build to another, such
as \href{https://genome.ucsc.edu/cgi-bin/hgLiftOver}{liftOver} from
UCSC.

\subsection{Aligners}\label{aligners}

\textbf{Aligners} are tools specifically designed to map reads to a
reference genome using different algorithms. Many aligners are
available, but the most used for aligning ChIP-seq data are:

\begin{itemize}
\tightlist
\item
  \textbf{\href{http://bowtie-bio.sourceforge.net/bowtie2/index.shtml}{Bowtie2}}.
  This is the one we will use in this workshop.
\item
  \textbf{\href{http://bio-bwa.sourceforge.net/}{BWA}}.
\end{itemize}

Both tools allow alignment of single and paired end reads, you just have
to specify it when running them. Generating an index is a techinque that
many aligners use to speed up their running type. You may think about it
as a book index: when having an idex it is then easier and faster to
find specific parts of the book (i.e.~the reference genome).
\href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2836519/}{Paper
describing indexing methods}

The output of the aligner is a \texttt{SAM/BAM} file containing the read
sequence and the assigned coordinates respect to the reference genome.
The aligner will report the number of aligned reads (which should be
\textgreater{}80\% of the sequenced reads) and which of those where
uniquely mapped or multi-mapping (they match more than one region in the
genome). Aligners such as bowtie2 report the best alginment for
multi-mapping reads.

\definecolor{unnamed-chunk-7.body.fill}{RGB}{173, 216, 230}
\definecolor{unnamed-chunk-7.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-7.header.fill}{RGB}{138, 173, 184}
\definecolor{unnamed-chunk-7.header.colour}{RGB}{0, 0, 0}

\cboxs[Pro-tip]{unnamed-chunk-7.body.fill!100}{unnamed-chunk-7.body.colour!100}{unnamed-chunk-7.header.fill!100}{unnamed-chunk-7.header.colour!100}

Many papers have tried to compare the performance of the different
aligners available. However, my advice to you is to look in the
literature for papers analyzing data similar to yours and just copy
their methods!

\cboxe

\subsection{\texorpdfstring{\texttt{SAM}/\texttt{BAM}
files}{SAM/BAM files}}\label{sambam-files}

\textbf{SAM} (Sequence Alignment/Map) and \textbf{BAM} (same as SAM but
binary, compressed and indexed) are file types used for storing sequence
and alignment information.

\definecolor{unnamed-chunk-8.body.fill}{RGB}{221, 160, 221}
\definecolor{unnamed-chunk-8.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-8.header.fill}{RGB}{177, 128, 177}
\definecolor{unnamed-chunk-8.header.colour}{RGB}{0, 0, 0}

\cboxs[Exercise]{unnamed-chunk-8.body.fill!100}{unnamed-chunk-8.body.colour!100}{unnamed-chunk-8.header.fill!100}{unnamed-chunk-8.header.colour!100}

Can you tell which one of the folloing files is in SAM and which one in
BAM format?

\emph{File 1}:

\begin{quote}
�BC�W{[}�Y�gbb\&J\A��``����έ��q�鮙4۷t�\\
��P�m\&=�m�z����
\textgreater{}+���ºB\(���|}�U�}[��U�._4�����T����������������/�WN?p�Bem��l�n��g��`4�.�'�����z������ �t��� !�I�\)��'1�M\#0��@\\
3�\textgreater{}�''�0��/æ�R��P�T�D
��L\&���\#�1BM������P2��p��`I\$�I\}�M�D\^{}t�P�h2f�cFɠ�ό�
\end{quote}

\emph{File 2}:

\begin{quote}
@HD VN:1.6 SO:coordinate\\
@SQ SN:ref LN:45\\
r001 99 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *\\
r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *\\
r003 0 ref 9 30 5S6M * 0 0 GCCTAAGCTAA * SA:Z:ref,29,-,6H5M,17,0;
\end{quote}

\cboxe

\definecolor{unnamed-chunk-9.body.fill}{RGB}{172, 255, 175}
\definecolor{unnamed-chunk-9.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-9.header.fill}{RGB}{138, 204, 140}
\definecolor{unnamed-chunk-9.header.colour}{RGB}{0, 0, 0}

\cboxs[Look at the result]{unnamed-chunk-9.body.fill!30}{unnamed-chunk-9.body.colour!100}{unnamed-chunk-9.header.fill!30}{unnamed-chunk-9.header.colour!100}

\begin{itemize}
\tightlist
\item
  File 1 is a BAM file. As BAM files are binary and compressed, they are
  not readable.
\item
  File 2 is a SAM file. SAM files are easily readable but their size is
  much bigger than BAM files!
\end{itemize}

\cboxe

BAM/SAM files consist on two different parts:

\begin{itemize}
\tightlist
\item
  \textbf{Header}: starts with \texttt{@XX} and contains information on
  the file itself: how it is ordered, which is the reference genome
  used, which program you used to align the reads, etc.
\item
  \textbf{Body}: contains information on the sequences, their genomic
  coordinates, mapping qualities, etc.
\end{itemize}

You can find more information on all the fields present in the SAM/BAM
files from the
\href{https://samtools.github.io/hts-specs/SAMv1.pdf}{Samtools
documentation}.

\definecolor{unnamed-chunk-10.body.fill}{RGB}{221, 160, 221}
\definecolor{unnamed-chunk-10.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-10.header.fill}{RGB}{177, 128, 177}
\definecolor{unnamed-chunk-10.header.colour}{RGB}{0, 0, 0}

\cboxs[Exercise]{unnamed-chunk-10.body.fill!100}{unnamed-chunk-10.body.colour!100}{unnamed-chunk-10.header.fill!100}{unnamed-chunk-10.header.colour!100}

Can you tell which lines are the header and which lines are the body of
this SAM file?

\begin{quote}
@HD VN:1.6 SO:coordinate\\
@SQ SN:ref LN:45\\
r001 99 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *\\
r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *\\
r003 0 ref 9 30 5S6M * 0 0 GCCTAAGCTAA * SA:Z:ref,29,-,6H5M,17,0;\\
r004 0 ref 16 30 6M14N5M * 0 0 ATAGCTTCAGC *\\
r003 2064 ref 29 17 6H5M * 0 0 TAGGC * SA:Z:ref,9,+,5S6M,30,1;\\
r001 147 ref 37 30 9M = 7 -39 CAGCGGCAT * NM:i:1
\end{quote}

\cboxe

\definecolor{unnamed-chunk-11.body.fill}{RGB}{172, 255, 175}
\definecolor{unnamed-chunk-11.body.colour}{RGB}{0, 0, 0}
\definecolor{unnamed-chunk-11.header.fill}{RGB}{138, 204, 140}
\definecolor{unnamed-chunk-11.header.colour}{RGB}{0, 0, 0}

\cboxs[Look at the result]{unnamed-chunk-11.body.fill!30}{unnamed-chunk-11.body.colour!100}{unnamed-chunk-11.header.fill!30}{unnamed-chunk-11.header.colour!100}

\begin{itemize}
\tightlist
\item
  The first two lines starting with \texttt{@HD} and \texttt{@SQ} are
  the header.

  \begin{itemize}
  \tightlist
  \item
    The first line (\texttt{HD}) is telling you the format version
    (\texttt{VN:1.6}) and that the file is sorted by coordinates
    (\texttt{SO:coordinate}).
  \item
    The second line (\texttt{SQ}) contains information on the reference
    genome used for the alignment: the name of the reference genome
    (\texttt{SN:ref}) and it length (\texttt{LN:45}).
  \end{itemize}
\item
  The following lines contain infomation the each one of the reads
  aligned to the reference genome:

  \begin{enumerate}
  \def\labelenumi{\arabic{enumi}.}
  \tightlist
  \item
    Read name.
  \item
    Flag. Bitwise information on the read status (mapped, unmapped,
    paired, unpaired, etc.).
  \item
    Reference sequence name.
  \item
    Position in the sequence
  \item
    Mapping quality (MAPQ)
  \item
    {[}\ldots{}{]}
  \end{enumerate}
\end{itemize}

{[}\ldots{}{]} You can find more information on all the fields present
in the SAM/BAM files from the
\href{https://samtools.github.io/hts-specs/SAMv1.pdf}{Samtools
documentation}.

\cboxe

\subsection{Hands on! Let's align this
thing!}\label{hands-on-lets-align-this-thing}

\subsubsection{Step 1: Download and index the reference
genome.}\label{step-1-download-and-index-the-reference-genome.}

Before running an alignment, you will have to download the reference
genome you want to use and \textbf{index} it.

For downloading hg19 you can go to
\href{http://hgdownload.cse.ucsc.edu/downloads.html}{UCSC downloads}
\textgreater{} Human \textgreater{} hg19 \textgreater{} Full data set
\textgreater{} \texttt{hg19.2bit}. \texttt{2bit} is a compression that
UCSC uses for making fasta files smaller. You can convert a
\texttt{2bit} file into a \texttt{fasta} file by running
\texttt{twoBitToFasta} from {[}UCSC
tools{]}((\url{http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/}).

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{gunzip}\NormalTok{ hg19.fa.gz }\CommentTok{# Uncompress the file}
\ExtensionTok{bowtie2-build}\NormalTok{ [options] }\OperatorTok{<}\NormalTok{reference_genome}\OperatorTok{>} \OperatorTok{<}\NormalTok{bt2_index_baseName}\OperatorTok{>} \CommentTok{# Create index with Bowtie2}
\end{Highlighting}
\end{Shaded}

\subsubsection{Step 2: Align your reads to the reference
genome.}\label{step-2-align-your-reads-to-the-reference-genome.}

We got our index and we got our reads, so let's run the alignment! You
can specify different parameters for the alignment (check
\texttt{bowtie2\ -h} to see them all), we are going to use the
following:

\begin{itemize}
\tightlist
\item
  \texttt{-t}: Print the time it takes to perform each step.
\item
  \texttt{-x}: Index filename prefix (without the trailing
  \texttt{.X.bt2}).
\item
  \texttt{-U}: Single-end reads file (if it were paired end you need to
  specify the first pair file with \texttt{-1} and the second pair with
  \texttt{-2}). Fastq files can be gzipped (extension: \texttt{.gz}), so
  no need to unzip before aligning!
\item
  \texttt{-S}: Specify file for SAM output.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{bowtie2}\NormalTok{ -t -x index -U single_end_reads.fastq -S output.sam}
\end{Highlighting}
\end{Shaded}

\subsubsection{Step 3: Explore the output of the
alignment}\label{step-3-explore-the-output-of-the-alignment}

\begin{itemize}
\tightlist
\item
  Percentage of alignment.
\item
  File size.
\item
  File format.
\end{itemize}

\section{Post-processing}\label{post-processing}

Once your alignment is done, you usually will have to perform some
routine steps to process the the output:

\begin{itemize}
\tightlist
\item
  The output is usually is a \texttt{SAM} file. You should convert it to
  \texttt{BAM}, which is a more light-weight way of storing the same
  data.
\item
  You should sort your \texttt{BAM} file by genomic coordinates.
\item
  You can also filter out some reads you do not want in your
  \texttt{BAM} file: duplicates, reads mapping to ENCODE black-listed
  regions, chromosomes you are not interested in, etc.
\item
  You should \textbf{index} your \texttt{BAM} file. As with reference
  genomes, you can create a index for your \texttt{BAM} file so
  operations requiring accessing parts of it will run much faster.
\end{itemize}

Usually, for running operations on \texttt{BAM} or \texttt{SAM} files,
the preferred program is \href{http://www.htslib.org/}{Samtools}.

\subsection{\texorpdfstring{Converting \texttt{SAM} to
\texttt{BAM}}{Converting SAM to BAM}}\label{converting-sam-to-bam}

This first step can be easily done using \texttt{samtools\ view}, with
the following arguments:

\begin{itemize}
\tightlist
\item
  \texttt{-@\ 4}: Indicates the number (4) of \textbf{additional}
  threads to use (default is 0).
\item
  \texttt{-b}: Tells samtools to convert to \texttt{BAM}.
\item
  \texttt{-o}: Output filename for the \texttt{BAM} file.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{samtools}\NormalTok{ view data/ChIP-seq/BAM/file.sam -@ 4 -b -o data/ChIP-seq/BAM/file.raw.bam}
\end{Highlighting}
\end{Shaded}

\subsection{\texorpdfstring{Sorting your \texttt{BAM} file by
coordinate}{Sorting your BAM file by coordinate}}\label{sorting-your-bam-file-by-coordinate}

\begin{itemize}
\tightlist
\item
  \texttt{-o}: Output filename for the sorted \texttt{BAM} file.
\item
  \texttt{-m\ 2G}: Tells Samtools to use up to 2G \textbf{per thread}.
\item
  \texttt{-@\ 4}: Uses 4 \_\_additional\_\_threads.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{samtools}\NormalTok{ sort data/ChIP-seq/BAM/file.raw.bam -o data/ChIP-seq/BAM/file.srtd.bam -m 2G -@ 4}
\end{Highlighting}
\end{Shaded}

\subsection{Removing duplicates}\label{removing-duplicates}

Could be an issue for small reads, single-end TF ChIP-seq\ldots{}
Histone marks are broader, so the probability of having a biologically
duplicated fragment is smaller.

Removing duplicates:

\begin{itemize}
\item
  \texttt{MarkDuplicates} from Picard.
\item
  \texttt{markdup} from Samtools.
\item
  \texttt{-r}: Remove duplicate reads.
\item
  \texttt{-@\ 4}: Use 4 \textbf{additional} threads.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{samtools}\NormalTok{ markdup data/ChIP-seq/BAM/file.srtd.bam data/ChIP-seq/BAM/file.rmdup.bam -r -@ 4}
\end{Highlighting}
\end{Shaded}

\subsection{Indexing your BAM file}\label{indexing-your-bam-file}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{samtools}\NormalTok{ index file.rmdup.bam}
\end{Highlighting}
\end{Shaded}

\section{Peak calling}\label{peak-calling}

\section{Visualization}\label{visualization}

\section{Additional resources}\label{additional-resources}

\begin{itemize}
\tightlist
\item
  \href{https://www.ebi.ac.uk/training/online/sites/ebi.ac.uk.training.online/files/user/18/private/chipseq_loos.pdf}{NGS
  ChIP-seq course, EMBL-EBI}
\item
  \href{https://github.com/crazyhottommy/ChIP-seq-analysis}{Resources
  for ChIP-seq analysis}
\item
  \href{http://biocluster.ucr.edu/~rkaundal/workshops/R_feb2016/ChIPseq/ChIPseq.html}{R/Bioconductor
  ChIP-seq workflow}
\item
  \href{https://github.com/shenlab-sinai/chip-seq_preprocess}{Pipeline
  for ChIP-seq processing}
\item
  \href{https://github.com/mforde84/ENCODE_TF_ChIP_pipeline}{ENCODE TF
  ChIP-seq pipeline}
\item
  \href{http://www.biologie.ens.fr/~mthomas/other/chip-seq-training/}{Hands-on
  introduction to ChIP-seq analysis - VIB Training}
\item
  \href{https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003326}{Practical
  guidelines for the comprehensive analysis of ChIP-seq data}
\end{itemize}


\end{document}
